### DISCUSSION TITLE
lasR for production?

### DISCUSSION THREAD

**Discussion by wiesehahn** [r-lidar/lasR#5] [General]
Hey, thanks for developing `lidR` and now this also seems to be promising. 
If I remember correctly I heard something like 
>lidR is meant for prototyping and LAStools should be used for production.

Now `lasR` seems very much geared towards effiency. What's the intended use-case for lasR once leaving early development stage? Would you still sign the above sentence?

**Comment by Jean-Romain** [comment]
I don't know where does the citation come from, but it could be a real one :wink: 

- `lasR` has less tool than `LAStools` but some of them have no equivalent in any other software on the market such as `rasterize()` or `callback()` that are capable of running user defined R functions.
- `lasR` is not at the level of efficiency of `LAStools`. I did not compare, but I'm pretty confident that `LAStools` remains by far the best in terms of speed, memory efficiency, robustness and so on. Actually I do not really care, I'm not in competition with anybody and I do not try to have a marketing position. I have nothing to sell.

The intended use cases are described more or less [here](https://r-lidar.github.io/lasR/articles/lasR1.html). In short:

- if you know exactly what you want and what you want can be done with `lasR`, do it with `lasR` especially for large coverages. For example, you know that you want a DTM and a raster of your custom metrics over 5000 km². Use `lasR`. This is very common and basic and `lasR` will do it better and faster
- if you want to explore, visualize, test, retry, visualize again to check if it works, retry again with improvements then use `lidR` it had been designed for that very purpose.

I'm not currently using `lasR` myself. I'm currently working on bridge segmentation. I spent my time applying algebra on clipped point clouds, looking at the results and especially at intermediate results to debug, test, retry, improve, validate and so on. I'm using `lidR` because it has much more tools and is designed for this very kind of usages.

In another project, I will have to compute area based metrics on 8000 km² of single photon lidar data. I'll use `lasR`.

`lidR` was designed to explore and design applications easily but is not a production software. It is exactly what LAStools is not. It was not designed to build large CHM and DTM but rather it has CHM and DTM tools to be able to build a CHM/DTM that might be useful in your exploration process and your developments of new methods.

Yet, many users are using `lidR` like if it were a production software. So I made `lasR`. It might be useful to me as well.

**Reply by wiesehahn** [reply]
Thanks for clarification. 

> I don't know where does the citation come from, but it could be a real one 😉

Either I read it somewhere or it might have been Alexis at last years SilviLaser

I calculated some indices using lidR for ~50.000 km² once, blocking the workstation for a couple of weeks in holidays. And I know people used it on a high performance cluster. So there is a way if you want, but lasR might come in handy.

**Reply by Jean-Romain** [reply]
50.000 km² will be blocking for a couple of weeks either way.

It is worth to notice that I developed `lasR` on a laptop that is 10 years old with 4 cores and 12 GB of RAM. I was usually not able to leverage parallelization with `future` on this laptop because `lidR` uses too much memory. But I recently changed my laptop and the benchmarks show that the differences between `lasR` and `lidR` are less significant. I'm still running some benchmark and tests. This is one of the reasons I consider `lasR` is still an experimental package.

As long as `lasR` does not support parallelization on multiple files, I think it can't beat `lidR` on a big ass computer with plenty of RAM. Yet, there are things that `lasR` can do and `lidR` can't.
